c++/src/capnp/schema.h

```diff
@@ -349,9 +462,14 @@ class StructSchema::Field {
   StructSchema parent;
   uint index;
   schema::Field::Reader proto;
+  Type type;
+
+  static Type resolveType(StructSchema parent, uint index, schema::Field::Reader proto);
 
   inline Field(StructSchema parent, uint index, schema::Field::Reader proto)
-      : parent(parent), index(index), proto(proto) {}
+      : parent(parent), index(index), proto(proto) {
+    type = resolveType(parent, index, proto);
```

Comment:
Unfortunately I don't think pushing resolution to here solves very much. Instances of `Field` are themselves created on-demand when accessing the `StructSchema` interface, e.g. when you iterate over a `FieldList` or call `getFieldByName()`. You can see there that the `Field` is returned by value, implying that a new `Field` is created each time.

Essentially all of the schema API types are actually created on-demand. The long-lived data structures underlying them are defined in `raw-schema.h`. In order to make a meaningful difference here I think we need to change the raw schemas to maintain appropriate lookup tables.

Reply:
> Before: 257ms, 44.5% of on-CPU time spent in getType().
> After: 151ms, 0.0% of on-CPU time spent in getType().

The second line here should really measure time spent in `resolveType()`, since all the compute from `getType()` was moved there. But you're reporting that the application did actually get faster overall?

I guess these numbers imply that your application is obtaining `Field` objects and then using them multiple times, such that the memoized type is actually useful?

So maybe the change is useful as-is. But, I'm sort of concerned that there may be applications which iterate fields but don't actually call `getType()` on them. Those apps will now be slower.

Reply:
> Unfortunately I don't think pushing resolution to here solves very much. Instances of `Field` are themselves created on-demand when accessing the `StructSchema` interface, e.g. when you iterate over a `FieldList` or call `getFieldByName()`. You can see there that the `Field` is returned by value, implying that a new `Field` is created each time.

The way it works for us internally to convert lua objects into capnproto:

1. We iterate the schema recursively and remember what each name corresponds to which `Field`. The `Field` objects are cached into a hashmap to avoid calling into `getFieldByName()` for every lua object we want to serialize.
2. Whenever an object is serialized, we find a corresponding `Field` _in our cache_ and reuse that. The `Field` objects are not created here, they are retrieved from a cache.

> Essentially all of the schema API types are actually created on-demand.

To reiterate: for us this is not the case. We create `Field` objects once when we discover the schema.

> The long-lived data structures underlying them are defined in `raw-schema.h`. In order to make a meaningful difference here I think we need to change the raw schemas to maintain appropriate lookup tables.

I _think_ this would make sense for a general case, as the speedup would be available for both the way we use it and the way you envision it to be used. I'm not sure if I'm capable of making that change though, but I'll see what I can do.

> The second line here should really measure time spent in `resolveType()`, since all the compute from `getType()` was moved there. But you're reporting that the application did actually get faster overall?

Hopefully this is clear by now from the previous paragraphs, but in our usage we create `Field` once and reuse it.

I don't think it would matter if we start measuring time before this line or after:

```c++
auto field = capnp::Type::from<Message>().asStruct().getFields()[0];
```

After all, we create `Field` once and run `set()` a million times, so the time is dominated by `set()`, which is confirmed by the width of the `set()` slab on the flamegraphs. Creating a `Field` and running `resolveType()` is but a blip in this.

The whole speedup is from not running `resolveType()` / original `getType()` a million times, but running it once.

> I guess these numbers imply that your application is obtaining Field objects and then using them multiple times, such that the memoized type is actually useful?

That is correct.

> So maybe the change is useful as-is. But, I'm sort of concerned that there may be applications which iterate fields but don't actually call `getType()` on them. Those apps will now be slower.

That is a reasonable concern to have. I'm not sure if it's a common or even a reasonable usage though.

Having all the caching in `raw-schema.h` would work for either case, but as I mentioned above it's harder for me to go that deep. I rarely touch C++ to begin with.